Mathlib/Algebra/Lie/Sl2.lean

```diff
@@ -0,0 +1,154 @@
+eigenvector for the action of both `h` and `e` necessarily has eigenvalue zero for `e`. -/
+lemma HasPrimitiveVectorWith.mk' [NoZeroSMulDivisors ℤ M] (t : IsSl2Triple h e f) (m : M) (μ ρ : R)
+    (hm : m ≠ 0) (hm' : ⁅h, m⁆ = μ • m) (he : ⁅e, m⁆ = ρ • m) :
+    HasPrimitiveVectorWith h e f m μ  where
+  h_ne_zero := t.h_ne_zero
+  lie_e_f := t.lie_e_f
+  lie_h_e_nsmul := t.lie_h_e_nsmul
+  lie_h_f_nsmul := t.lie_h_f_nsmul
+  ne_zero := hm
+  lie_h := hm'
+  lie_e := by
+    suffices 2 • ⁅e, m⁆ = 0 by simpa using this
+    rw [← nsmul_lie, ← t.lie_h_e_nsmul, lie_lie, hm', lie_smul, he, lie_smul, hm',
+      smul_smul, smul_smul, mul_comm ρ μ, sub_self]
+
+namespace HasPrimitiveVectorWith
+
+variable {m : M} {μ : R} (P : HasPrimitiveVectorWith h e f m μ)
+
+local notation "ψ" n => ((toEndomorphism R L M f) ^ n) m
+
+lemma lie_h_pow_toEndomorphism_f (n : ℕ) :
+    ⁅h, ψ n⁆ = (μ - 2 * n) • ψ n := by
+  induction' n with n ih
+  · simpa using P.lie_h
+  · rw [pow_succ', LinearMap.mul_apply, toEndomorphism_apply_apply, Nat.cast_add, Nat.cast_one,
+      leibniz_lie h, P.lie_lie_smul_f R, ← neg_smul, ih, lie_smul, smul_lie, ← add_smul]
+    congr
+    ring
+
+lemma lie_f_pow_toEndomorphism_f (n : ℕ) :
+    ⁅f, ψ n⁆ = ψ (n + 1) := by
+  simp [pow_succ']
+
+lemma lie_e_pow_succ_toEndomorphism_f (n : ℕ) :
+    ⁅e, ψ (n + 1)⁆ = ((n + 1) * (μ - n)) • ψ n := by
+  induction' n with n ih
+  · simp only [zero_add, pow_one, toEndomorphism_apply_apply, Nat.cast_zero, sub_zero, one_mul,
+      pow_zero, LinearMap.one_apply, leibniz_lie e, P.lie_e_f, P.lie_e, P.lie_h, lie_zero,
+      add_zero]
+  · rw [pow_succ', LinearMap.mul_apply, toEndomorphism_apply_apply, leibniz_lie e, P.lie_e_f,
+      lie_h_pow_toEndomorphism_f P, ih, lie_smul, lie_f_pow_toEndomorphism_f, ← add_smul,
+      Nat.cast_add, Nat.cast_one]
+    congr
+    ring
+
+/-- The eigenvalue of a primitive vector must be a natural number if the representation is
+finite-dimensional. -/
+lemma exists_nat [IsNoetherian R M] [NoZeroSMulDivisors R M] [IsDomain R] [CharZero R] :
+    ∃ n : ℕ, μ = n := by
+  suffices ∃ n : ℕ, (ψ n) = 0 by
+    obtain ⟨n, hn₁, hn₂⟩ := Nat.exists_not_and_succ_of_not_zero_of_exists P.ne_zero this
+    refine ⟨n, ?_⟩
+    have := lie_e_pow_succ_toEndomorphism_f P n
+    rw [hn₂, lie_zero, eq_comm, smul_eq_zero_iff_left hn₁, mul_eq_zero, sub_eq_zero] at this
+    exact_mod_cast this.resolve_left <| Nat.cast_add_one_ne_zero n
+  have hs : (Set.range <| fun (n : ℕ) ↦ μ - 2 * n).Infinite := by
+    rw [← Set.image_univ, Set.infinite_image_iff (Function.Injective.injOn (fun n m ↦ by simp) _)]
+    exact Set.infinite_univ
+  by_contra! contra
+  exact hs ((toEndomorphism R L M h).eigenvectors_linearIndependent
+    {μ - 2 * n | n : ℕ}
+    (fun ⟨s, hs⟩ ↦ ψ Classical.choose hs)
+    (fun ⟨s, hs⟩ ↦ by simp [lie_h_pow_toEndomorphism_f P, Classical.choose_spec hs, contra,
+      Module.End.HasEigenvector, Module.End.mem_eigenspace_iff])).finite
```

Comment:
This term is quite complicated... also you are shadowing `hs` (the head symbol of the term) inside the proof in the `fun ⟨s, hs⟩` clauses... maybe use `r, hr` there?

Reply:
Thanks, done (in #13076)